File: node/src/repos.ts
import { mkdir, rm, writeFile } from "node:fs/promises";
import path from "node:path";
import { cacheDir } from "./config.js";
import { GitHubClient } from "./github.js";
import { remoteFingerprint } from "./hash.js";
import type { SyncProgressReporter } from "./progress.js";
import { parseGithubRepo } from "./repoParse.js";

export interface RepoConfig {
  slug: string;
  owner: string;
  repo: string;
  branch: string;
  skillsPath: string;
}

export interface RemoteSkill {
  name: string;
  remotePath: string;
  fingerprint: string;
  files: Record<string, string>;
}

export function repoFromConfig(slug: string, data: any): RepoConfig {
  const [owner, repoName] = parseGithubRepo(data.owner, data.repo);
  return {
    slug,
    owner,
    repo: repoName,
    branch: data.branch ?? "main",
    skillsPath: data.skills_path ?? "skills",
  };
}

export function repoCacheDir(repo: RepoConfig): string {
  return path.join(cacheDir(), repo.slug);
}

function parseSkillTrees(tree: any[], skillsPath: string): Record<string, Record<string, string>> {
  const prefix = `${skillsPath.replace(/\/$/, "")}/`;
  const skills: Record<string, Record<string, string>> = {};
  for (const item of tree) {
    if (item.type !== "blob") continue;
    const itemPath = item.path as string;
    if (!itemPath.startsWith(prefix)) continue;
    const remainder = itemPath.slice(prefix.length);
    const slash = remainder.indexOf("/");
    if (slash === -1) continue;
    const skillName = remainder.slice(0, slash);
    const relPath = remainder.slice(slash + 1);
    skills[skillName] ??= {};
    skills[skillName][relPath] = item.sha;
  }
  return skills;
}

export async function fetchRemoteInventory(
  client: GitHubClient,
  repo: RepoConfig,
): Promise<[string, Record<string, RemoteSkill>]> {
  const commitSha = await client.resolveCommitSha(repo.owner, repo.repo, repo.branch);
  const tree = await client.fetchTree(repo.owner, repo.repo, repo.branch);
  const bySkill = parseSkillTrees(tree, repo.skillsPath);
  const inventory: Record<string, RemoteSkill> = {};
  for (const [name, files] of Object.entries(bySkill).sort()) {
    if (!files["SKILL.md"]) continue;
    inventory[name] = {
      name,
      remotePath: `${repo.skillsPath}/${name}`,
      fingerprint: remoteFingerprint(files),
      files,
    };
  }
  return [commitSha, inventory];
}

export async function downloadSkillFiles(
  client: GitHubClient,
  repo: RepoConfig,
  skill: RemoteSkill,
  dest: string,
  progress?: SyncProgressReporter,
): Promise<void> {
  await rm(dest, { recursive: true, force: true });
  await mkdir(dest, { recursive: true });
  const files = Object.entries(skill.files).sort(([a], [b]) => a.localeCompare(b));
  for (let index = 0; index < files.length; index += 1) {
    const [relPath, blobSha] = files[index];
    progress?.downloading(skill.name, relPath, index + 1, files.length);
    const content = await client.downloadBlob(repo.owner, repo.repo, blobSha);
    const target = path.join(dest, relPath);
    await mkdir(path.dirname(target), { recursive: true });
    await writeFile(target, content);
  }
}

File: python/src/skills_utils/repos.py
from __future__ import annotations

from dataclasses import dataclass
from pathlib import Path
from typing import Any

from skills_utils.config_loader import cache_dir
from skills_utils.github import GitHubClient
from skills_utils.hash import remote_fingerprint
from skills_utils.progress import SyncProgress
from skills_utils.repo_parse import parse_github_repo


@dataclass
class RepoConfig:
    slug: str
    owner: str
    repo: str
    branch: str
    skills_path: str

    @property
    def full_name(self) -> str:
        return f"{self.owner}/{self.repo}"


@dataclass
class RemoteSkill:
    name: str
    remote_path: str
    fingerprint: str
    files: dict[str, str]


def repo_from_config(slug: str, data: dict[str, Any]) -> RepoConfig:
    owner, repo_name = parse_github_repo(data["owner"], data["repo"])
    return RepoConfig(
        slug=slug,
        owner=owner,
        repo=repo_name,
        branch=data.get("branch", "main"),
        skills_path=data.get("skills_path", "skills"),
    )


def repo_cache_dir(repo: RepoConfig) -> Path:
    return cache_dir() / repo.slug


def parse_skill_trees(
    tree: list[dict[str, Any]], skills_path: str
) -> dict[str, dict[str, str]]:
    prefix = f"{skills_path.strip('/')}/"
    skills: dict[str, dict[str, str]] = {}
    for item in tree:
        if item.get("type") != "blob":
            continue
        path = item.get("path", "")
        if not path.startswith(prefix):
            continue
        remainder = path[len(prefix) :]
        parts = remainder.split("/", 1)
        if len(parts) != 2:
            continue
        skill_name, rel_path = parts
        skills.setdefault(skill_name, {})[rel_path] = item["sha"]
    return skills


def fetch_remote_inventory(
    client: GitHubClient, repo: RepoConfig
) -> tuple[str, dict[str, RemoteSkill]]:
    commit_sha = client.resolve_commit_sha(repo.owner, repo.repo, repo.branch)
    tree = client.fetch_tree(repo.owner, repo.repo, repo.branch)
    by_skill = parse_skill_trees(tree, repo.skills_path)
    inventory: dict[str, RemoteSkill] = {}
    for name, files in sorted(by_skill.items()):
        if "SKILL.md" not in files:
            continue
        inventory[name] = RemoteSkill(
            name=name,
            remote_path=f"{repo.skills_path}/{name}",
            fingerprint=remote_fingerprint(files),
            files=files,
        )
    return commit_sha, inventory


def download_skill_files(
    client: GitHubClient,
    repo: RepoConfig,
    skill: RemoteSkill,
    dest: Path,
    progress: SyncProgress | None = None,
) -> None:
    if dest.exists():
        import shutil

        shutil.rmtree(dest)
    dest.mkdir(parents=True, exist_ok=True)
    files = sorted(skill.files.items())
    for index, (rel_path, blob_sha) in enumerate(files, start=1):
        if progress:
            progress.downloading(skill.name, rel_path, index, len(files))
        content = client.download_blob(repo.owner, repo.repo, blob_sha)
        target = dest / rel_path
        target.parent.mkdir(parents=True, exist_ok=True)
        target.write_bytes(content)
